File: src/main/meeting-manager.ts
import { getDb } from './db/database'
import { setSetting } from './db/settings'
import type { BusyBlock } from '@shared/types'
import { log } from './logger'

const BUSY_UNTIL_KEY = 'busy_until'

export function getBusyUntil(): string | null {
  const db = getDb()
  const row = db.prepare('SELECT value FROM settings WHERE key = ?').get(BUSY_UNTIL_KEY) as
    | { value: string }
    | undefined
  return row?.value || null
}

export function setBusyUntil(ts: string): void {
  setSetting(BUSY_UNTIL_KEY, ts)
}

export function clearBusyUntil(): void {
  const db = getDb()
  db.prepare('DELETE FROM settings WHERE key = ?').run(BUSY_UNTIL_KEY)
}

function findMeetingsCategoryId(): number | null {
  const db = getDb()
  const row = db.prepare("SELECT id FROM categories WHERE name = 'Meetings' AND is_active = 1").get() as
    | { id: number }
    | undefined
  return row?.id ?? null
}

export function createMeetingEntries(busyBlock: BusyBlock): void {
  const categoryId = findMeetingsCategoryId()
  if (!categoryId) {
    log.error('No active "Meetings" category found — skipping meeting entries')
    return
  }

  const db = getDb()
  const startMs = new Date(busyBlock.start).getTime()
  const endMs = new Date(busyBlock.end).getTime()
  const durationMinutes = Math.round((endMs - startMs) / 60_000)

  db.prepare(
    `INSERT INTO entries (category_id, prompted_at, responded_at, credited_minutes)
     VALUES (?, ?, ?, ?)`
  ).run(categoryId, busyBlock.start, busyBlock.start, durationMinutes)
}

File: Sources/Services/MeetingManager.swift
import Foundation

final class MeetingManager {
    static let shared = MeetingManager()
    private init() {}

    func createMeetingEntries(busyBlock: BusyBlock) {
        // Find the "Meetings" category, then find or create an "unspecified" activity for it
        let categories = CategoryStore.shared.listCategories()
        guard let meetingsCat = categories.first(where: { $0.name == "Meetings" }),
              let categoryId = meetingsCat.id else {
            Logger.shared.error("No active \"Meetings\" category found — skipping meeting entries")
            return
        }

        guard let activity = ActivityStore.shared.findOrCreate(name: "unspecified", categoryId: categoryId),
              let activityId = activity.id else {
            Logger.shared.error("Failed to find/create meeting activity")
            return
        }

        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]

        guard let startDate = formatter.date(from: busyBlock.start),
              let endDate = formatter.date(from: busyBlock.end) else {
            // Try without fractional seconds
            formatter.formatOptions = [.withInternetDateTime]
            guard let startDate = formatter.date(from: busyBlock.start),
                  let endDate = formatter.date(from: busyBlock.end) else {
                Logger.shared.error("Failed to parse meeting dates")
                return
            }
            let durationMinutes = endDate.timeIntervalSince(startDate) / 60
            EntryStore.shared.createEntry(
                activityId: activityId,
                promptedAt: busyBlock.end,
                respondedAt: busyBlock.end,
                creditedMinutes: durationMinutes.rounded()
            )
            return
        }

        let durationMinutes = endDate.timeIntervalSince(startDate) / 60
        EntryStore.shared.createEntry(
            activityId: activityId,
            promptedAt: busyBlock.end,
            respondedAt: busyBlock.end,
            creditedMinutes: durationMinutes.rounded()
        )
    }
}
